src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -36,19 +38,107 @@ internal static bool IsEnabled()
+            {
+                activity = new Activity(DiagnosticsHandlerLoggingStrings.ActivityName);
+
+                s_diagnosticListener.StartActivity(activity, new { Request = request });
+
+                //inject headers unless user does not want it to happen
+                //we may pass request to IsEnabled, but not clear how useful it is
+                if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ActivityPropagateName))
+                {
+                    request.Headers.Add(HttpKnownHeaderNames.RequestId, activity.Id);
+                    
+                    //we expect baggage to be empty or contain a few items
+                    bool baggageIsEmpty = true;
+                    foreach (var _ in activity.Baggage)
+                    {
+                        baggageIsEmpty = false;
+                        break;
+                    }
+
+                    if (!baggageIsEmpty)
+                    {
+                        List<string> baggage = new List<string>();
+                        foreach (KeyValuePair<string, string> baggageItem in activity.Baggage)
+                        {
+                            baggage.Add(new NameValueHeaderValue(baggageItem.Key, baggageItem.Value).ToString());
+                        }
+                        request.Headers.Add(HttpKnownHeaderNames.CorrelationContext, baggage);
+                    }
+                }
+            }
+
+            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
+            try
+            {
+                await responseTask.ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                //we'll report task status in Activity.Stop
+                throw;
+            }
+            catch (Exception ex)
+            {
+                //If request was initialy instrumented, Activity.Current has all necessary context for logging
+                //If user decided to NOT instrument this request AND it threw an exception then:
+                //Activity.Current represents 'parent' Activity (presumably incoming request)
+                //So we let user log it as exception happened in this 'parent' activity
+                if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ExceptionEventName))
+                {
+                    s_diagnosticListener.Write(DiagnosticsHandlerLoggingStrings.ExceptionEventName, new {Exception = ex});
+                }
+                throw;
+            }
+            finally
+            {
+                //always stop activity if it was started
+                if (activity != null)
+                {
+                    s_diagnosticListener.StopActivity(activity, new
+                    {
+                        Response = responseTask.Status == TaskStatus.RanToCompletion ? responseTask.Result : null,
```

Comment:
"Request" and a possible "Exception" might be interesting here as well. This would allow subscribers to subscribe just to the Stop-Event and get all scenarios/information.